blog: Securing MCP (#1953)
block/goose: documentation/blog/2025-03-31-securing-mcp/index.md

---
title: "Securing the Model Context Protocol"

> description: Before you plug your agent into just any MCP server, here's how to check if it's actually safe.

Author: ```suggestion
description: Building secure and capable AI integrations with Model Context Protocol (MCP) at Block.
```

Reviewer: oops i didnt see that.

### Agent, Human and Device Identity

In our traditional model of Authentication (AuthN) and Authorization (AuthZ) it’s common to tie an identity to a single point of abstraction such as a person or a business.

This field has organically been evolving towards pairing a services identity user identity abstraction with identification of client devices such as browsers and mobile phones. This is done to help reduce the prevalence of attacks caused by automation and inauthentic traffic such as account takeover attacks (ATO).

With the evolution of Agents performing actions on behalf of users we will need to evolve to be able to determine the combination of:

1. The primary identity abstraction
2. The agent’s identity
3. The device/location the agent is running from 

Having consistent mechanisms for identifying usage in this manner allows companies to protect users from integrations with malicious agents and protect their platforms from attacks by unwanted agentic tooling.

> The model context protocol itself has a specification for [OAuth that at the time of writing is a draft](https://spec.modelcontextprotocol.io/specification/draft/basic/authorization/). 

Reviewer 2: They killed this link 😢 
```suggestion
The model context protocol itself has a specification for [OAuth that at the time of writing was a draft, but has since been released here](https://spec.modelcontextprotocol.io/specification/2025-03-26/basic/authorization/). 
```

1. Only install MCPs from trusted sources and are well maintained
2. Implement integrity checks and/or signing of artifacts where possible to ensure you’re executing the expected code
3. Implement allow lists on enterprise agents to ensure users only use pre-validated MCPs

## Conclusion

Much like agents are paving the way to allow LLMs to have more real-world utility MCP and similar protocols will continue to grow in adoption. 

We believe that by contributing to open source projects early, sharing our learnings publicly, and building our own solutions that leverage MCP, Block can maintain security best practices from the deterministic world while continuing to evolve them with newer technologies.

We’re excited to work on making this protocol more secure for users and developers alike and are looking forward to sharing how we’ve used MCP for our own Security use-cases in the future.

<head>
  <meta property="og:title" content="Securing the Model Context Protocol" />
  <meta property="og:type" content="article" />
  <meta property="og:url" content="https://block.github.io/goose/blog/2025/03/31/securing-mcp" />

>   <meta property="og:description" content="Before you plug your AI agent into just any MCP server, here's how to check if it's actually safe." />

Author: ```suggestion
  <meta property="og:description" content="Building secure and capable AI integrations with Model Context Protocol (MCP) at Block." />
```

>   <meta property="og:image" content="http://block.github.io/goose/assets/images/mcpsafety-87eb7ace7163a5edbe068ff75b79a199.png" />

Author: ```suggestion
  <meta property="og:image" content="http://block.github.io/goose/assets/images/securing-mcp-5e475e91c0e621afa33e30b3d89ef065.png" />
```

  <meta name="twitter:card" content="summary_large_image" />
  <meta property="twitter:domain" content="block.github.io/goose" />
  <meta name="twitter:title" content="Securing the Model Context Protocol" />

>   <meta name="twitter:description" content="Before you plug your agent into just any MCP server, here's how to check if it's actually safe." />

Author: ```suggestion
  <meta name="twitter:description" content="Building secure and capable AI integrations with Model Context Protocol (MCP) at Block." />
```

>   <meta name="twitter:image" content="http://block.github.io/goose/assets/images/mcpsafety-87eb7ace7163a5edbe068ff75b79a199.png" />

Author: ```suggestion
  <meta name="twitter:image" content="http://block.github.io/goose/assets/images/securing-mcp-5e475e91c0e621afa33e30b3d89ef065.png" />
```
